runagent.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 qbee.io
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+)
+
+// RunAgentManager is a manager for running the agent on devices.
+type RunAgentManager struct {
+	// client is the RunAgentManager client.
+	client *Client
+	// allowFailures allows failures.
+	allowFailures bool
+}
+
+// NewRunAgentManager returns a new RunAgentManager.
+func NewRunAgentManager() *RunAgentManager {
+	return &RunAgentManager{}
+}
+
+// WithClient sets the client
+func (m *RunAgentManager) WithClient(client *Client) *RunAgentManager {
+	m.client = client
+	return m
+}
+
+// WithAllowFailures sets the allowFailures flag
+func (m *RunAgentManager) WithAllowFailures(allow bool) *RunAgentManager {
+	m.allowFailures = allow
+	return m
+}
+
+// RunAgentDevices is a response for the run agent devices request.
+type RunAgentDevices struct {
+	Items []struct {
+		NodeID string `json:"node_id"`
+	} `json:"items"`
+}
+
+// RunAgentDevice runs the agent on the device.
+func (m *RunAgentManager) RunAgentDevice(ctx context.Context, deviceID string) error {
+	err := m.runAgent(ctx, deviceID)
+	if err == nil {
+		fmt.Printf("Successfully triggered agent on device %s\n", deviceID)
+		return nil
+	}
+	if m.allowFailures {
+		fmt.Printf("Failed to run agent on device %s: %s\n", deviceID, err.Error())
+		return nil
+	}
+	return fmt.Errorf("failed to run agent on device %s: %w", deviceID, err)
+}
+
+// RunAgentGroup runs the agent on all devices in the group.
+func (m *RunAgentManager) RunAgentGroup(ctx context.Context, groupID string) error {
+	path := fmt.Sprintf("/api/v2/inventory?search={\"ancestor_ids\":[\"%s\"]}", groupID)
+
+	var devices RunAgentDevices
+
+	if err := m.client.Call(ctx, http.MethodGet, path, nil, &devices); err != nil {
+		return err
+	}
+
+	return m.runAgentMultiple(ctx, devices)
+
+}
+
+// RunAgentTag runs the agent on all devices in the group with the specified tag.
+func (m *RunAgentManager) RunAgentTag(ctx context.Context, tag string) error {
+	path := fmt.Sprintf("/api/v2/inventory?search={\"tags\":[\"%s\"]}", tag)
```

Comment:
We have `ListDeviceInventory` to search for devices, I would suggest we use that instead.